text/0000-reads-clauses-on-methods.md

```diff
@@ -0,0 +1,200 @@
+
+# Summary
+[summary]: #summary
+
+This proposal simply adds support for optional `reads` clauses on methods, just as they are already supported on functions. While the default for functions is disallowing all reads (i.e. `reads {}`) the default for methods would be allowing all reads (i.e. `reads *`), allowing full backwards-compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+Dafny currently has no direct support for concurrency. The soundness of method verification depends on the assumption that no modification to mutable state occurs other than what is present in the method body. Consider this example of a Dafny class with internal state:
+
+```dafny
+class Counter {
+  var x: int
+  constructor() {
+    x := 0;
+  }
+  function method Get() 
+    reads this
+    ensures Get() == x
+  {
+    x
+  }
+  method Increment() 
+    modifies this
+    ensures x == old(x) + 1
+  {
+    x := x + 1;
+  }
+}
+```
+
+If this class is compiled to C#, for example, and its instances are shared between multiple threads, its verified properties are no longer true. If two threads both invoke `Increment` on a single counter, the effect could easily be to only increment `x` once, since reading and writing `x` does not occur atomically. Even the post-condition on `Get` is no longer valid, because the value of `x` could change in-between reading its value and returning from the compiled body.
+
+This proposal does not aim to add concurrency primitives to Dafny, or even propose any particular choice of paradigm for asynchronous or concurrent execution. Instead, it supports verifying that a particular method is safe to use in a concurrent environment, by verifying that it does not operate on any mutable state unsafely.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In most Dafny code, you will not have to specify a reads clause on any methods, as they are permitted to read any mutable state by default. In some cases the specification for a method needs to be more precise, however. In particular, invoking a compiled method from external, native code in a concurrent environment is much more likely to be safe if the method does not read or write any shared mutable state. It will now be possible to specify this requirement by using `reads` clauses on methods as well as functions. For example:
+
+```dafny
+
+class FibonacciMemoizer {
+  var cache: map<nat, nat>
+  constructor() reads {} {
+    cache := map[];
+  }
+  method Get(n: nat) returns (r: nat) 
+    reads this
+    modifies this
+  {
+    if n <= 2 {
+      return 1;
+    }
+    if n in cache {
+      r := cache[n];
+    } else {
+      var oneBefore := Get(n - 1);
+      var twoBefore := Get(n - 2);
+      r := oneBefore + twoBefore;
+      cache := cache[n := r];
+    }
+  }
+}
+
+// Sharing a memoizer instance between threads is not allowed.
+method {:extern} UnsafeFibonacci(n: nat, memoizer: FibonacciMemoizer) returns (r: nat)
```

Comment:
It's definitely possible because the meaning of `{:extern}` is unfortunately overloaded: it can means the method is either externally implemented or externally called. We've had some discussion about improving that in the past: https://github.com/dafny-lang/dafny/issues/469

I'll add comments to make it clearer though.

Reply:
Ah, thanks :)